### ISSUE TITLE
Revisiting Feature Alignment for One-stage Object Detection

### ISSUE THREAD

**Issue by runzeer** [TuSimple/simpledet#209]
when will the source codes for the AlignDet be released?

**Comment by RogerChern** [comment]
Thanks for the interset. The code will be released once the paper is accepted.

**Comment by boundles** [comment]
您好，我是这样实现的adm模块，训练的时候loss经常为nan，麻烦帮看下有问题吗？

`import math
import torch
import torch.nn.functional as F
from torch import nn

from .inference import  make_aligndet_postprocessor
from .loss import make_aligndet_adm_loss_evaluator
from ..anchor_generator import make_anchor_generator_aligndet

from lib.modeling.box_coder import BoxCoder
from lib.layers import DeformConv

def normal_init(module, mean=0, std=1, bias=0):
    nn.init.normal_(module.weight, mean, std)
    if hasattr(module, 'bias') and module.bias is not None:
        nn.init.constant_(module.bias, bias)

class RoiConv(nn.Module):
    """Feature Alignment Module.
    Feature Alignment Module is implemented based on DCN v1.
    It uses anchor shape prediction rather than feature map to
    predict offsets of deformable conv layer.
    Args:
        in_channels (int): Number of channels in the input feature map.
        out_channels (int): Number of channels in the output feature map.
        kernel_size (int): Deformable conv kernel size.
        deformable_groups (int): Deformable conv group size.
    """

    def __init__(self,
                 in_channels,
                 out_channels,
                 kernel_size=3,
                 deformable_groups=4):
        super(RoiConv, self).__init__()
        offset_channels = kernel_size * kernel_size * 2

        self.offset_nums = int(offset_channels / 2 * deformable_groups)
        self.conv_offset = nn.Conv2d(4,
                                     deformable_groups * offset_channels,
                                     1,
                                     bias=False)
        self.conv_adaption = DeformConv(in_channels,
                                        out_channels,
                                        kernel_size=kernel_size,
                                        padding=(kernel_size - 1) // 2,
                                        deformable_groups=deformable_groups)
        self.relu = nn.ReLU(inplace=True)

    def init_weights(self):
        normal_init(self.conv_offset, std=0.1)
        normal_init(self.conv_adaption, std=0.01)

    def gen_feature_loc(self, x):
        N, _, H, W = x.shape
        device = x.device
        loc_x = torch.arange(
            0, W, step=1, dtype=torch.float32, device=device
        )
        loc_y = torch.arange(
            0, H, step=1, dtype=torch.float32, device=device
        )
        mesh_x, mesh_y = torch.meshgrid((loc_x, loc_y))
        loc = torch.stack((mesh_x, mesh_y), dim=0)
        loc = loc.permute(0, 2, 1)
        loc = loc.repeat(N, self.offset_nums, 1, 1)
        return loc

    def forward(self, x, box_loc, stride):
        box_loc = box_loc / stride 
        feature_loc = self.gen_feature_loc(x)
        offset = self.conv_offset(box_loc) + feature_loc
        x = self.relu(self.conv_adaption(x, offset))
        return x

class AlignDetADM(nn.Module):
    """
    AlignDet Dense Proposal Module
    """
    def __init__(self, cfg, in_channels):
        """
        Arguments:
            in_channels (int): number of channels of the input feature
            num_anchors (int): number of anchors to be predicted
        """
        super(AlignDetADM, self).__init__()
        # TODO: Implement the sigmoid version first.
        num_classes = cfg.MODEL.ALIGNDET.NUM_CLASSES - 1
        num_anchors = len(cfg.MODEL.ALIGNDET.ASPECT_RATIOS)
        
        self.anchor_strides = cfg.MODEL.ALIGNDET.ANCHOR_STRIDES

        self.roi_conv = RoiConv(in_channels, in_channels * 4, kernel_size=3)

        self.mid_conv = nn.Conv2d(
            in_channels * 4, in_channels * 4, kernel_size=1, bias=False
        )

        self.cls_logits = nn.Conv2d(
            in_channels * 4, num_anchors * num_classes, kernel_size=3, stride=1,
            padding=1
        )
        self.bbox_pred = nn.Conv2d(
            in_channels * 4,  num_anchors * 4, kernel_size=3, stride=1,
            padding=1
        )

        # Initialization
        for modules in [self.cls_logits, self.bbox_pred]:
            for l in modules.modules():
                if isinstance(l, nn.Conv2d):
                    torch.nn.init.normal_(l.weight, std=0.01)
                    torch.nn.init.constant_(l.bias, 0)

        self.roi_conv.init_weights()

        normal_init(self.mid_conv, std=0.01)

        # aligndet_bias_init
        prior_prob = cfg.MODEL.ALIGNDET.PRIOR_PROB
        bias_value = -math.log((1 - prior_prob) / prior_prob)
        torch.nn.init.constant_(self.cls_logits.bias, bias_value)

    def forward(self, x, box_out_init):
        logits = []
        bbox_reg = []
        for feature, per_box_out_init, stride in zip(x, box_out_init, self.anchor_strides):
            aligned_feature = self.roi_conv(feature, per_box_out_init, stride)
            aligned_feature = self.mid_conv(aligned_feature)
            logits.append(self.cls_logits(aligned_feature))
            bbox_reg.append(self.bbox_pred(aligned_feature))
            
        return logits, bbox_reg


class AlignDetADMModule(torch.nn.Module):
    """
    ADM Module for AlignDet computation. Takes feature maps from the backbone and
    AlignDet outputs and losses. Only Test on FPN now.
    """

    def __init__(self, cfg):
        super(AlignDetADMModule, self).__init__()

        self.cfg = cfg.clone()
        in_channels = cfg.MODEL.BACKBONE.OUT_CHANNELS
        anchor_generator = make_anchor_generator_aligndet(cfg)
        head = AlignDetADM(cfg, in_channels)
        box_coder= BoxCoder(weights=(10., 10., 5., 5.))
        
        box_selector_test = make_aligndet_postprocessor(cfg, box_coder, is_train=False)

        loss_evaluator = make_aligndet_adm_loss_evaluator(cfg, box_coder)
            
        if self.cfg.MODEL.MASK_ON:
            self.box_selector_train = make_aligndet_postprocessor(cfg, box_coder, is_train=True)

        self.anchor_generator = anchor_generator
        self.head = head
        self.box_selector_test = box_selector_test
        self.loss_evaluator = loss_evaluator

    def forward(self, image_sizes, features, box_loc_init, targets=None):
        """
        Arguments:
            image_sizes (list[Tensor]): sizes of images
            features (list[Tensor]): features computed from the images that are
                used for computing the predictions. Each tensor in the list
                correspond to different feature levels
            targets (list[BoxList): ground-truth boxes present in the image (optional)
        Returns:
            boxes (list[BoxList]): the predicted boxes from the RPN, one BoxList per
                image.
            losses (dict[Tensor]): the losses for the model during training. During
                testing, it is an empty dict.
        """
        box_cls, box_regression = self.head(features, box_loc_init)
        anchors = self.anchor_generator(image_sizes, features)
 
        if self.training:
            return self._forward_train(anchors, box_cls, box_regression, targets)
        else:
            # only support single scale
            boxes = []
            sizes = []
            anchors = list(zip(*anchors))
            for boxlists in anchors:
                box = [boxlist.bbox for boxlist in boxlists]
                box = torch.cat(box)
                boxes.append(box)
                sizes.append(torch.Tensor([boxlists[0].size[0], boxlists[0].size[1]]))
            return self._forward_test(boxes, sizes, box_cls, box_regression)

    def _forward_train(self, anchors, box_cls, box_regression, targets):

        loss_box_cls, loss_box_reg = self.loss_evaluator(
            anchors, box_cls, box_regression, targets
        )

        losses = {
            "loss_aligndet_cls_adm": loss_box_cls,
            "loss_aligndet_reg_adm": loss_box_reg
        }

        if self.cfg.MODEL.MASK_ON:
            with torch.no_grad():
                boxes = self.box_selector_train(anchors, box_cls, box_regression, targets)
        else:
            boxes = anchors
        
        return boxes, losses

    def _forward_test(self, boxes, im_sizes, box_cls, box_regression):
        boxes = self.box_selector_test(boxes, im_sizes, box_cls, box_regression)
        return boxes

def build_aligndet_adm(cfg):
    return AlignDetADMModule(cfg)`

**Comment by boundles** [comment]
> Thanks for the interset. The code will be released once the paper is accepted.

您好，我是这么实现的adm，但训练的时候经常出现nan，麻烦帮看下实现上是有问题吗，谢谢。

`import math
import torch
import torch.nn.functional as F
from torch import nn

from .inference import  make_aligndet_postprocessor
from .loss import make_aligndet_adm_loss_evaluator
from ..anchor_generator import make_anchor_generator_aligndet

from lib.modeling.box_coder import BoxCoder
from lib.layers import DeformConv

def normal_init(module, mean=0, std=1, bias=0):
    nn.init.normal_(module.weight, mean, std)
    if hasattr(module, 'bias') and module.bias is not None:
        nn.init.constant_(module.bias, bias)

class RoiConv(nn.Module):
    """Feature Alignment Module.
    Feature Alignment Module is implemented based on DCN v1.
    It uses anchor shape prediction rather than feature map to
    predict offsets of deformable conv layer.
    Args:
        in_channels (int): Number of channels in the input feature map.
        out_channels (int): Number of channels in the output feature map.
        kernel_size (int): Deformable conv kernel size.
        deformable_groups (int): Deformable conv group size.
    """

    def __init__(self,
                 in_channels,
                 out_channels,
                 kernel_size=3,
                 deformable_groups=4):
        super(RoiConv, self).__init__()
        offset_channels = kernel_size * kernel_size * 2

        self.offset_nums = int(offset_channels / 2 * deformable_groups)
        self.conv_offset = nn.Conv2d(4,
                                     deformable_groups * offset_channels,
                                     1,
                                     bias=False)
        self.conv_adaption = DeformConv(in_channels,
                                        out_channels,
                                        kernel_size=kernel_size,
                                        padding=(kernel_size - 1) // 2,
                                        deformable_groups=deformable_groups)
        self.relu = nn.ReLU(inplace=True)

    def init_weights(self):
        normal_init(self.conv_offset, std=0.1)
        normal_init(self.conv_adaption, std=0.01)

    def gen_feature_loc(self, x):
        N, _, H, W = x.shape
        device = x.device
        loc_x = torch.arange(
            0, W, step=1, dtype=torch.float32, device=device
        )
        loc_y = torch.arange(
            0, H, step=1, dtype=torch.float32, device=device
        )
        mesh_x, mesh_y = torch.meshgrid((loc_x, loc_y))
        loc = torch.stack((mesh_x, mesh_y), dim=0)
        loc = loc.permute(0, 2, 1)
        loc = loc.repeat(N, self.offset_nums, 1, 1)
        return loc

    def forward(self, x, box_loc, stride):
        box_loc = box_loc / stride 
        feature_loc = self.gen_feature_loc(x)
        offset = self.conv_offset(box_loc) + feature_loc
        x = self.relu(self.conv_adaption(x, offset))
        return x

class AlignDetADM(nn.Module):
    """
    AlignDet Dense Proposal Module
    """
    def __init__(self, cfg, in_channels):
        """
        Arguments:
            in_channels (int): number of channels of the input feature
            num_anchors (int): number of anchors to be predicted
        """
        super(AlignDetADM, self).__init__()
        # TODO: Implement the sigmoid version first.
        num_classes = cfg.MODEL.ALIGNDET.NUM_CLASSES - 1
        num_anchors = len(cfg.MODEL.ALIGNDET.ASPECT_RATIOS)
        
        self.anchor_strides = cfg.MODEL.ALIGNDET.ANCHOR_STRIDES

        self.roi_conv = RoiConv(in_channels, in_channels * 4, kernel_size=3)

        self.mid_conv = nn.Conv2d(
            in_channels * 4, in_channels * 4, kernel_size=1, bias=False
        )

        self.cls_logits = nn.Conv2d(
            in_channels * 4, num_anchors * num_classes, kernel_size=3, stride=1,
            padding=1
        )
        self.bbox_pred = nn.Conv2d(
            in_channels * 4,  num_anchors * 4, kernel_size=3, stride=1,
            padding=1
        )

        # Initialization
        for modules in [self.cls_logits, self.bbox_pred]:
            for l in modules.modules():
                if isinstance(l, nn.Conv2d):
                    torch.nn.init.normal_(l.weight, std=0.01)
                    torch.nn.init.constant_(l.bias, 0)

        self.roi_conv.init_weights()

        normal_init(self.mid_conv, std=0.01)

        # aligndet_bias_init
        prior_prob = cfg.MODEL.ALIGNDET.PRIOR_PROB
        bias_value = -math.log((1 - prior_prob) / prior_prob)
        torch.nn.init.constant_(self.cls_logits.bias, bias_value)

    def forward(self, x, box_out_init):
        logits = []
        bbox_reg = []
        for feature, per_box_out_init, stride in zip(x, box_out_init, self.anchor_strides):
            aligned_feature = self.roi_conv(feature, per_box_out_init, stride)
            aligned_feature = self.mid_conv(aligned_feature)
            logits.append(self.cls_logits(aligned_feature))
            bbox_reg.append(self.bbox_pred(aligned_feature))
            
        return logits, bbox_reg


class AlignDetADMModule(torch.nn.Module):
    """
    ADM Module for AlignDet computation. Takes feature maps from the backbone and
    AlignDet outputs and losses. Only Test on FPN now.
    """

    def __init__(self, cfg):
        super(AlignDetADMModule, self).__init__()

        self.cfg = cfg.clone()
        in_channels = cfg.MODEL.BACKBONE.OUT_CHANNELS
        anchor_generator = make_anchor_generator_aligndet(cfg)
        head = AlignDetADM(cfg, in_channels)
        box_coder= BoxCoder(weights=(10., 10., 5., 5.))
        
        box_selector_test = make_aligndet_postprocessor(cfg, box_coder, is_train=False)

        loss_evaluator = make_aligndet_adm_loss_evaluator(cfg, box_coder)
            
        if self.cfg.MODEL.MASK_ON:
            self.box_selector_train = make_aligndet_postprocessor(cfg, box_coder, is_train=True)

        self.anchor_generator = anchor_generator
        self.head = head
        self.box_selector_test = box_selector_test
        self.loss_evaluator = loss_evaluator

    def forward(self, image_sizes, features, box_loc_init, targets=None):
        """
        Arguments:
            image_sizes (list[Tensor]): sizes of images
            features (list[Tensor]): features computed from the images that are
                used for computing the predictions. Each tensor in the list
                correspond to different feature levels
            targets (list[BoxList): ground-truth boxes present in the image (optional)
        Returns:
            boxes (list[BoxList]): the predicted boxes from the RPN, one BoxList per
                image.
            losses (dict[Tensor]): the losses for the model during training. During
                testing, it is an empty dict.
        """
        box_cls, box_regression = self.head(features, box_loc_init)
        anchors = self.anchor_generator(image_sizes, features)
 
        if self.training:
            return self._forward_train(anchors, box_cls, box_regression, targets)
        else:
            # only support single scale
            boxes = []
            sizes = []
            anchors = list(zip(*anchors))
            for boxlists in anchors:
                box = [boxlist.bbox for boxlist in boxlists]
                box = torch.cat(box)
                boxes.append(box)
                sizes.append(torch.Tensor([boxlists[0].size[0], boxlists[0].size[1]]))
            return self._forward_test(boxes, sizes, box_cls, box_regression)

    def _forward_train(self, anchors, box_cls, box_regression, targets):

        loss_box_cls, loss_box_reg = self.loss_evaluator(
            anchors, box_cls, box_regression, targets
        )

        losses = {
            "loss_aligndet_cls_adm": loss_box_cls,
            "loss_aligndet_reg_adm": loss_box_reg
        }

        if self.cfg.MODEL.MASK_ON:
            with torch.no_grad():
                boxes = self.box_selector_train(anchors, box_cls, box_regression, targets)
        else:
            boxes = anchors
        
        return boxes, losses

    def _forward_test(self, boxes, im_sizes, box_cls, box_regression):
        boxes = self.box_selector_test(boxes, im_sizes, box_cls, box_regression)
        return boxes

def build_aligndet_adm(cfg):
    return AlignDetADMModule(cfg)`

**Comment by RogerChern** [comment]
@boundles 
You need to figure out the exact layout of offsets used your implementation of DeformConv. And you can try to opt out the RoIConv for vanilla convolution to see if the network converges.

**Comment by boundles** [comment]
> @boundles
> You need to figure out the exact layout of offsets used your implementation of DeformConv. And you can try to opt out the RoIConv for vanilla convolution to see if the network converges.

再咨询个问题，ADM模块的梯度会回传到DPM吗？
